cirq/algorithms/qaoa/QAOA.py

```diff
@@ -0,0 +1,353 @@
+        :param min_flag: bool that determines whether to return negative of cost
+
+        :return: expectation value of self.objective_function in state prepared
+            by circuit created using params
+        """
+
+        # clears old circuit
+        self.circuit = Circuit()
+
+        # new circuit to prepare QAOA state with appropriate params
+        self.setup_circuit(p, params)
+
+        # calculates expectation value in state according to method
+        if expectation_method == 'wavefunction':
+            cost = expectation_value(circuit=self.circuit,
+                                     operator=self.objective_function)
+
+        elif isinstance(expectation_method, int):
+            cost = expectation_from_sampling(circuit=self.circuit,
+                                             operator=self.objective_function,
+                                             n_samples=expectation_method)
+
+        else:
+            raise TypeError('method must be wavefunction or an '
+                            'integer representing number of samples')
+
+        # default is maximizing objectvie function, if flag is passed,
+        # multiplies it by -1
+        if min_flag:
+            return -1*cost
+
+        return cost
+
+    def optimization(self, p,
+                     initial_params=None,
+                     optimizer='swarm',
+                     expectation_method='wavefunction'):
+        """
+        Main function that optimizes the parameters for QAOA
+
+        :param p: number of repetitions of the unitaries U_C and U_B used in
+            the state preparation in QAOA. Since each time a unitary is applied
+            we introduce a new parameter, there are 2*p total parameters we
+            want to optimize
+        :param initial_params:
+        :param optimizer: string that sets the optimization method used to
+            obtain the parameters gamma and beta that maximize the objective
+            function. Currently supports:
+            swarm - swarming algorithm
+            NM - Nelder-Mead
+            bayesian - Bayesian optimization
+        :param expectation_method: determines whether expectation value
+            is obtained by an inner product with the final state or
+            by sampling from measurements given repeated runs from the circuit
+            if sampling is desired, enter int corresponding to the number of
+            samples used to get expectation value
+
+        :return: returns a dictionary of the solution found using the given
+            optimization method
+            keys:
+            cost - final value of the objective function we are maximizing
+            gammas - list containing the p optimal gammas
+            betas - list containing the p  optimal gammas
+        """
+
+        # initialize solution dict
+        solution = {}
+
+        if optimizer == 'swarm':
+
+            # lower and upper bounds for parameters in swarming optimization
```

Comment:
Split each of these optimizer cases into separate private methods that are called by this method.